opm/ml/keras_model.cpp

```diff
@@ -0,0 +1,419 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "keras_model.hpp"
+
+#include <cmath>
+#include <fstream>
+#include <limits>
+#include <stdio.h>
+#include <utility>
+
+#include <iostream>
+
+namespace Opm {
+
+
+bool ReadUnsignedInt(std::ifstream* file, unsigned int* i) {
+    KASSERT(file, "Invalid file stream");
+    KASSERT(i, "Invalid pointer");
+
+    file->read((char*)i, sizeof(unsigned int));
+    KASSERT(file->gcount() == sizeof(unsigned int), "Expected unsigned int");
+
+    return true;
+}
+
+bool ReadFloat(std::ifstream* file, float* f) {
+    KASSERT(file, "Invalid file stream");
+    KASSERT(f, "Invalid pointer");
+
+    file->read((char*)f, sizeof(float));
+    KASSERT(file->gcount() == sizeof(float), "Expected Evaluation");
+
+    return true;
+}
+
+bool ReadFloats(std::ifstream* file, float* f, size_t n) {
+    KASSERT(file, "Invalid file stream");
+    KASSERT(f, "Invalid pointer");
+
+    file->read((char*)f, sizeof(float) * n);
+    KASSERT(((unsigned int)file->gcount()) == sizeof(float) * n,
+            "Expected Evaluations");
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerActivation<Evaluation>::LoadLayer(std::ifstream* file) {
+    KASSERT(file, "Invalid file stream");
+
+    unsigned int activation = 0;
+    KASSERT(ReadUnsignedInt(file, &activation),
+            "Failed to read activation type");
+
+    switch (activation) {
+    case kLinear:
+        activation_type_ = kLinear;
+        break;
+    case kRelu:
+        activation_type_ = kRelu;
+        break;
+    case kSoftPlus:
+        activation_type_ = kSoftPlus;
+        break;
+    case kHardSigmoid:
+        activation_type_ = kHardSigmoid;
+        break;
+    case kSigmoid:
+        activation_type_ = kSigmoid;
+        break;
+    case kTanh:
+        activation_type_ = kTanh;
+        break;
+    default:
+        KASSERT(false, "Unsupported activation type %d", activation);
+    }
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerActivation<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    KASSERT(in, "Invalid input");
+    KASSERT(out, "Invalid output");
+
+    *out = *in;
+
+    switch (activation_type_) {
+    case kLinear:
+        break;
+    case kRelu:
+        for (size_t i = 0; i < out->data_.size(); i++) {
+            if (out->data_[i] < 0.0) {
+                out->data_[i] = 0.0;
+            }
+        }
+        break;
+    case kSoftPlus:
+        for (size_t i = 0; i < out->data_.size(); i++) {
+            out->data_[i] = log(1.0 + exp(out->data_[i]));
+        }
+        break;
+    case kHardSigmoid:
+        for (size_t i = 0; i < out->data_.size(); i++) {
+            Evaluation x = (out->data_[i] * 0.2) + 0.5;
```

Comment:
If these constant are universal and there is no need to ever change them, give them a name and define them:
```
constexpr double sigmoid_scale = 0.2;
...
```
Otherwise, they should probably be data members, and this particular activation function be a separate subclass of KerasLayer.